### ISSUE TITLE
Persistent Low R² and Unstable Pearson Correlation with MCF7 Data

### ISSUE THREAD

**Issue by faith-ogun** [GET-Foundation/get_model#14]
### **Subject:** Persistent Low R² and Unstable Pearson Correlation with MCF7 Data

Hi Dr Fu,

First of all, thank you for the amazing work on the **GET (General Expression Transformer) model**. I’ve been working on my project, which focuses on predicting gene expression for **MCF7 breast cancer cells** using both **ATAC-seq** and **RNA-seq** data for my final year dissertation. However, I’ve been facing persistent issues with the model’s performance, and I’m hoping to get some guidance.

### **Problem Summary:**
- **R² values** remain consistently **negative**, often close to **-1**.
- **Spearman correlation** is **around 0.3**, much lower than expected.
- **Validation loss (`val_loss`) and expression loss (`val_exp_loss`) appear as flat lines at zero** throughout training.
- **Variance warning during training**:  
  *"The variance of predictions or target is close to zero. This can cause instability in Pearson correlation coefficient."*

---

### **My Setup:**
- **ATAC-seq Data**: ENCODE **IDR thresholded peaks** and pseudoreplicated peaks.
- **RNA-seq Data**: Normalized **aTPM using expected counts**, following the tutorial’s preprocessing steps.
- **Chromatin Accessibility Window**: 300bp around the TSS (as per tutorial recommendations).
- **Finetuning Checkpoint**: Initially, I used the checkpoint from `s3://2023-get-xf2217/get_demo/checkpoints/regulatory_inference_checkpoint_fetal_adult/finetune_fetal_adult_leaveout_astrocyte/checkpoint-best.pt` but have also tried without a checkpoint.

---

### **What I’ve Tried So Far:**
1. **Tested different types of peaks**—using both **pseudoreplicated peaks** and **IDR threshold peaks**. Unfortunately, the performance did not improve significantly.
2. **Increased the TSS window beyond 300bp** to capture broader enhancer activity. This led to slight improvements but did not fully resolve the issue.
3. **Finetuned the model from scratch** on MCF7 data rather than using the fetal-adult checkpoint. Despite the effort, the metrics remained far from acceptable.
4. **Double-checked preprocessing steps** for both ATAC-seq and RNA-seq data to ensure consistent normalization. The aTPM normalization follows the tutorial approach and has been verified.

---

Any suggestions or insights you could provide would be greatly appreciated! I’m happy to provide additional details if needed.

Thank you so much for your time and help!

Faith Ogundimu
Genetics and Cell Biology 4
Dublin City University
https://www.linkedin.com/in/faith-ogundimu-3895421b9/

_Originally posted by @faith-ogun in https://github.com/GET-Foundation/get_model/discussions/13_

**Comment by fuxialexander** [comment]
Just to clarify, what do you mean by RNA-seq aTPM value? Is it just TPM from RNA-seq data? If it is, is the data bulk or single cell? From my previous experience, bulk RNA-seq data does not have high R^2 with UMI based data (e.g. single cell RNA-seq). 
Have you tried to predict ATAC itself from motif? If that does not work there is probably some order issue between peak annotation and stored motif values.

**Comment by faith-ogun** [comment]
Yes, it's just the TPM from RNA-Seq data that has been normalized per the tutorial instructions.
The data is bulk RNA-seq from ENCODE using a gene quantifications tsv file and getting the counts and normalizing it.

I've tried to predict ATAC itself from the motif but when using the mac configuration, I get this error:
{'machine': {'data_path': '/Users/xf2217/Repos/get_model/', 'codebase': '/Users/xf2217/Repos/get_model/', 'output_dir': '/Users/xf2217/Downloads', 'num_devices': 1, 'num_workers': 4, 'batch_size': 4}}

I've already edited the code to specify my directories, but they come back as read-only when they aren't. Still trying to debug that.

I'm using a singularity container inside of a remote server which may be the issue?

**Comment by fuxialexander** [comment]
Did you do 
```
cfg.machine.data_path="/your/path"
```
to specify the directories?

Please do that after the following lines
```
from get_model.config.config import MachineConfig

cfg.machine = MachineConfig(**load_config('machine/mac').machine)
```
Otherwise they might be overwritten.

If you can find some scRNA-seq based MCF7 I would use those. The scale of scRNA and bulk are just not the same. When I do a Bulk ATAC->Bulk RNA recently I got ~0.56 spearman and around 0 R^2.

**Comment by faith-ogun** [comment]
Thank you I found RNA-Seq data and used that and got similar results! 

But why such a low R^2 ? Is it because I'm using cancer data?

**Comment by fuxialexander** [comment]
Rereading your issue above:
```
Validation loss (val_loss) and expression loss (val_exp_loss) appear as flat lines at zero throughout training.
```
This looks very weird. How can the `val_loss` be zero if the performance is not good?

I had a look at your jupyter notebooks in your repo and notice that you are using IDR peaks. Those are very robust by conservative peak calls. Please use a less stringent cut off for peaks. Alternatively, use the peak set in the bottom of https://github.com/GET-Foundation/get_model/issues/6 and similar code I provide their to get a larger and more consistent with our pretraining peak set and use the BAM file from MCF7 to recalculate the aTPM for all peaks.

**Comment by fuxialexander** [comment]
I found a single cell multiome data for MCF7 on GEO. Consider use that instead: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154873

**Comment by faith-ogun** [comment]
I will try that data, thank you! 💪

**Comment by faith-ogun** [comment]
Hi, thanks for all your help!

I ran it with the data you suggested, using only the controls with no estrogen treatment. However I'm still seeing the same issue.

Could it be the motif file and would I have to make my own that is MCF-7 specific?

i.e. this code from the preprocessing tutorial:
```python
motif_bed_url = "https://resources.altius.org/~jvierstra/projects/motif-clustering/releases/v1.0/hg38.archetype_motifs.v1.0.bed.gz"
motif_bed_index_url = "https://resources.altius.org/~jvierstra/projects/motif-clustering/releases/v1.0/hg38.archetype_motifs.v1.0.bed.gz.tbi"


if (
    motif_bed_url
    and motif_bed_index_url
    and not (
        (annotation_dir / "hg38.archetype_motifs.v1.0.bed.gz").exists()
        or (annotation_dir / "hg38.archetype_motifs.v1.0.bed.gz.tbi").exists()
    )
):
    download_motif(motif_bed_url, motif_bed_index_url, motif_dir=annotation_dir)
    motif_bed = str(annotation_dir / "hg38.archetype_motifs.v1.0.bed.gz")
else:
    motif_bed = str(annotation_dir / "hg38.archetype_motifs.v1.0.bed.gz")
```

I've attached an image of my run results at epoch 3.

Also,

 is there a way to do the pretrain myself, so it's trained on my data and then do the finetune with my own checkpoint rather than using the astrocyte/pbmc one?

<img width="1078" alt="Image" src="https://github.com/user-attachments/assets/9247f0a5-f5ec-4dd9-bdb4-bc410e863e29" />

**Comment by fuxialexander** [comment]
can you share the `atac.bed` and `rna.csv` or ideally the zarr file for troubleshooting? Maybe through google drive?

**Comment by faith-ogun** [comment]
Yes of course:
https://drive.google.com/drive/folders/18Ku-rWBR_kg7Ngdw5t51YY7f_01OJ5m2?usp=sharing

Zarr folder is still uploading but should be ready in a few minutes.

**Comment by fuxialexander** [comment]
Just got a time to run on your data. The ATAC prediction gives me ~0.75 correlation which means the motif & ATAC values are fine. I got similar (0.4) correlation on your expression data. When I take a closer look at the data, I found that the scale of most of your expression on autosome is around 0-2. Usually they should be 0-4 in log10 TPM. Upon closer look, I found that in your rna.csv file, you included mitochondrial genes which are extremely highly expressed (>4), they should be filtered out when you computing the TPM (usually a single cell analysis pipeline will automatically remove them, so I forgot to mention this step in the tutorial).  Please recompute the TPM without those genes and have another try.

**Comment by faith-ogun** [comment]
Awesome I'll give that a try! How did you do the correlation analysis for each sequencing dataset separately?

**Comment by fuxialexander** [comment]
Can you clarify the question a bit? which correlation analysis are you specifically mentioning?

**Comment by faith-ogun** [comment]
Yes of course, I can clarify, how did you get the .75 and .4? 

It would be nice to be able to run these tests before running the model, possibly as an exploratory measure and so I can do troubleshooting myself on a future dataset if necessary.

**Comment by fuxialexander** [comment]
0.75 is when running the `predict_atac.ipynb` tutorial with your data (which trains a motif->atac model). ~0.4 is when running the `finetune_pmbc.ipynb` tutorial with your data (a motif+atac -> expression model).

**Comment by faith-ogun** [comment]
I removed the mitochrondrial genes (MT-), but am still getting a really bad r^2 (~-0.6-1) with both the `finetune_pmbc.ipynb` and `finetune_pmbc_lo.ipynb`. 

I do not have access to GPU, only CPU.

I also uninstalled and reinstalled the model using the singularity instructions. Really not sure what could be going wrong. Surely I thought it may be my input but we're using the same one.

These are my parameters:
```python
cfg = load_config('finetune_tutorial_pbmc') # load the predefined finetune tutorial config
cfg.stage = 'fit'

# Set a unique project name for training on all chromosomes
cfg.run.project_name = 'finetune_multiome_control'
cfg.run.run_name = 'training_from_finetune_lora_chr10_split_QATAC_MCF7_no_mito_ATAC_fix'

cfg.dataset.quantitative_atac = False # We use binary ATAC signal for motif interpretation analysis

cfg.dataset.zarr_path = "/project/home/p200469/get_BIO1018/get_preprocess_output_multiome_control.zarr/"
cfg.dataset.celltypes = "all_chrs"

cfg.finetune.checkpoint = "./checkpoint-best.pth" # set the path to the pretrained checkpoint we want to finetune from
cfg.dataset.leave_out_celltypes = '' # set the celltypes you want to leave out, '' here means no celltype leave out
cfg.dataset.leave_out_chromosomes = 'chr10' # set the chromosomes you want to leave out, '' here means no chromosome leave out
cfg.machine.num_devices=0 # use 0 for cpu training; >=1 for gpu training
cfg.machine.batch_size=8 # batch size for training

cfg.machine.output_dir = "/project/home/p200469/get_BIO1018/get_ML_output"
cfg.training.epochs = 30

cfg.training.val_check_interval = 5.0 

print(f"output path: {cfg.machine.output_dir}/{cfg.run.project_name}/{cfg.run.run_name}")
print(f"training for {cfg.training.epochs} epochs")
```

Any place I could possibly be going wrong? Very strange to be getting a 

Attaching plot images and google drive link:

<img width="1058" alt="Image" src="https://github.com/user-attachments/assets/0aa9f723-c366-45e5-9adf-45ea9849985d" />

https://drive.google.com/drive/folders/18Ku-rWBR_kg7Ngdw5t51YY7f_01OJ5m2?usp=sharing

Thanks for your help throughout this!

**Comment by fuxialexander** [comment]
That's indeed very strange. I guess either it's due to low RNA-seq depth/quality or it's due to MCF7's highly messed up genome 

![Image](https://github.com/user-attachments/assets/3684d4de-7daa-4c49-9deb-a9083d6f6779)
